pcap/slicer.go

```diff
@@ -0,0 +1,105 @@
+package pcap
+
+import (
+	"encoding/json"
+	"errors"
+	"io"
+	"io/ioutil"
+	"os"
+
+	"github.com/brimsec/zq/pkg/nano"
+	"github.com/brimsec/zq/pkg/ranger"
+)
+
+// Slicer implements io.Reader reading the sliced regions provided to it from
+// the underlying file thus extracting subsets of an underlying file without
+// modifying or copying the file.
+type Slicer struct {
+	slices []Slice
+	file   *os.File
+	reader *io.SectionReader
+}
+
+func NewSlicer(file *os.File, index *Index, span nano.Span) (*Slicer, error) {
+	slices, err := GenerateSlices(index, span)
+	if err != nil {
+		return nil, err
+	}
+	first := slices[0]
+	return &Slicer{
+		slices: slices[1:],
+		file:   file,
+		reader: first.NewReader(file),
+	}, nil
+}
+
+func (s *Slicer) Read(b []byte) (int, error) {
+	for s.reader != nil {
+		n, err := s.reader.Read(b)
+		if n != 0 {
+			if err == io.EOF {
+				err = nil
+			}
+			return n, err
+		}
+		if err != io.EOF {
+			return 0, err
+		}
+		if len(s.slices) != 0 {
+			s.reader = s.slices[0].NewReader(s.file)
+			s.slices = s.slices[1:]
+		} else {
+			s.reader = nil
+		}
+	}
+	return 0, io.EOF
+}
+
+func (s Slice) Overlaps(x Slice) bool {
+	return x.Offset >= s.Offset && x.Offset < s.Offset+x.Length
+}
+
+func (s Slice) NewReader(r io.ReaderAt) *io.SectionReader {
+	return io.NewSectionReader(r, int64(s.Offset), int64(s.Length))
+}
```

Comment:
I would keep these grouped with the `Slice` type declaration, probably moving the type declaration here.

Reply:
ok